src/components/shared/thread/ActivityGraph.js

```diff
@@ -0,0 +1,322 @@
+  +samplesSelectedStates?: boolean[],
+  +treeOrderSampleComparator?: (
+    IndexIntoSamplesTable,
+    IndexIntoSamplesTable
+  ) => number,
+|};
+
+class ThreadActivityGraph extends PureComponent<ActivityGraphProps> {
+  _canvas: null | HTMLCanvasElement = null;
+  _resizeListener = () => this.forceUpdate();
+  _categoryDrawStyles: null | CategoryDrawStyle[] = null;
+  _activityGraphFills: null | ActivityGraphFills = null;
+
+  _takeCanvasRef = (canvas: HTMLCanvasElement | null) => {
+    this._canvas = canvas;
+  };
+
+  _renderCanvas() {
+    const canvas = this._canvas;
+    if (canvas !== null) {
+      timeCode('ThreadActivityGraph render', () => {
+        this.drawCanvas(canvas);
+      });
+    }
+  }
+
+  componentDidMount() {
+    window.addEventListener('resize', this._resizeListener);
+    this.forceUpdate(); // for initial size
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this._resizeListener);
+  }
+
+  /**
+   * Get or lazily create the category info. It requires the 2d ctx to exist in order
+   * to create the fill patterns.
+   */
+  _getCategoryDrawStyles(ctx: CanvasRenderingContext2D): CategoryDrawStyle[] {
+    if (this._categoryDrawStyles === null) {
+      // Lazily initialize this list.
+      this._categoryDrawStyles = this.props.categories.map(
+        ({ color: colorName }, categoryIndex) => {
+          const styles = _mapColorNameToStyles(colorName);
+          return {
+            ...styles,
+            category: categoryIndex,
+            filteredOutFillStyle: _createDiagonalStripePattern(
+              ctx,
+              styles.unselectedFillStyle
+            ),
+          };
+        }
+      );
+    }
+
+    return this._categoryDrawStyles;
+  }
+
+  drawCanvas(canvas: HTMLCanvasElement) {
+    const { fullThread } = this.props;
+    const { samples } = fullThread;
+
+    if (samples.length === 0) {
+      // Do not attempt to render when there are no samples.
+      return;
+    }
+    const rect = canvas.getBoundingClientRect();
+    const ctx = canvas.getContext('2d');
+    const canvasPixelWidth = Math.round(rect.width * window.devicePixelRatio);
+    const canvasPixelHeight = Math.round(rect.height * window.devicePixelRatio);
+    canvas.width = canvasPixelWidth;
+    canvas.height = canvasPixelHeight;
+
+    this._activityGraphFills = new ActivityGraphFills(
+      canvasPixelWidth,
+      canvasPixelHeight,
+      this.props,
+      this._getCategoryDrawStyles(ctx)
+    );
+
+    const fillResult = this._activityGraphFills.computeFills();
+    let { lastCumulativeArray } = fillResult;
+    const { categoryFills } = fillResult;
+
+    // Draw adjacent filled paths using Operator ADD and disjoint paths.
+    // This avoids any bleeding and seams.
+    // lighter === OP_ADD
+    ctx.globalCompositeOperation = 'lighter';
+    lastCumulativeArray = new Float32Array(canvasPixelWidth);
+    for (const { fillStyle, perPixelContribution } of categoryFills) {
```

Comment:
This loop could use a bit more documentation. For example: What's the meaning of `lastCumulativeArray` (it keeps track of where the "mountain ridge" is after the last fill), and what the deal with the non-zero ranges is. (Some fills might not span the full width of the graph - they have parts where their contribution stays zero for some time. So we split the fill calls into multiple, for each range during which the fill has an uninterrupted range of non-zero-contribution pixels.)